revm-utils/src/evm.rs

```diff
@@ -0,0 +1,228 @@
+                    .collect::<Vec<_>>()
+                    .into(),
+                gas_used: U256::from(result.gas_used()),
+                error: matches!(result, ExecutionResult::Revert { .. })
+                    .then_some(result.output().unwrap().to_string()),
+            },
+            accounts,
+        })
+    }
+
+    async fn call_inner(
+        &mut self,
+        tx: &N::TransactionRequest,
+        validate_tx: bool,
+    ) -> Result<(ExecutionResult, std::collections::HashMap<Address, Account>), EvmError> {
+        let mut db = ProofDB::new(self.tag, self.execution.clone());
+        _ = db.state.prefetch_state(tx, validate_tx).await;
+
+        let env_with_handler_cfg = self.get_env_with_handler_cfg(tx, validate_tx).await;
+        let evm = Revm::builder()
+            .with_db(db)
+            .with_env_with_handler_cfg(env_with_handler_cfg)
+            .build();
+
+        let mut ctx = evm.into_context_with_handler_cfg();
+
+        let tx_res = loop {
+            let db = ctx.context.evm.db.borrow_mut();
+            if db.state.needs_update() {
+                db.state.update_state().await.unwrap();
+            }
+
+            let mut evm = Revm::builder().with_context_with_handler_cfg(ctx).build();
+            let res = evm.transact();
+            ctx = evm.into_context_with_handler_cfg();
+
+            let db = ctx.context.evm.db.borrow_mut();
+            let needs_update = db.state.needs_update();
+
+            if res.is_ok() || !needs_update {
+                break res.map(|res| (res.result, std::mem::take(&mut db.state.accounts)));
+            }
+        };
+
+        tx_res.map_err(|err| EvmError::Generic(format!("generic: {}", err)))
+    }
+
+    async fn get_env_with_handler_cfg(
+        &self,
+        tx: &N::TransactionRequest,
+        validate_tx: bool,
+    ) -> EnvWithHandlerCfg {
+        let block = self
+            .execution
+            .get_block(self.tag.into(), false)
+            .await
+            .unwrap()
+            .ok_or(ExecutionError::BlockNotFound(self.tag))
+            .unwrap();
+
+        let spec_id = get_spec_id_for_block(
+            &self.network,
+            &self.fork_schedule,
+            block.header().timestamp(),
+        );
+        let spec_id = spec_id.unwrap_or(SpecId::LATEST);
+        let handler_cfg = HandlerCfg {
+            spec_id,
+            is_optimism: false,
+        };
+
+        let mut cfg = CfgEnv::default();
+        cfg.chain_id = self.chain_id;
+        cfg.disable_block_gas_limit = !validate_tx;
+        cfg.disable_eip3607 = !validate_tx;
+        cfg.disable_base_fee = !validate_tx;
+
+        let env = Env {
+            tx: Self::tx_env(tx),
+            block: Self::block_env(&block, &self.fork_schedule),
+            cfg,
+        };
+
+        EnvWithHandlerCfg::new(Box::new(env), handler_cfg)
+    }
+
+    fn tx_env(tx: &N::TransactionRequest) -> TxEnv {
+        // TODO: Handle the Tx fields based on networks
```

Comment:
We could always move back to generating this in the `NetworkSpec` where we have the required type information. I do think if we can figure out a solution that doesn't require that though it would be preferable.

Reply:
Kept the fields `None` and left a TODO to be handled later